File: clients/sellingpartner-api-aa-java/tst/com/amazon/SellingPartnerAPIAA/LWAAuthorizationSignerTest.java
package com.amazon.SellingPartnerAPIAA;

import com.squareup.okhttp.Call;
import com.squareup.okhttp.MediaType;
import com.squareup.okhttp.OkHttpClient;
import com.squareup.okhttp.Protocol;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;
import com.squareup.okhttp.ResponseBody;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.Arguments;
import org.junit.jupiter.params.provider.MethodSource;
import org.mockito.ArgumentCaptor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;

import java.io.IOException;
import java.util.Arrays;
import java.util.HashSet;
import java.util.stream.Stream;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertNotSame;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

import static com.amazon.SellingPartnerAPIAA.ScopeConstants.SCOPE_NOTIFICATIONS_API;
import static com.amazon.SellingPartnerAPIAA.ScopeConstants.SCOPE_MIGRATION_API;

public class LWAAuthorizationSignerTest {
    private static final String TEST_REFRESH_TOKEN = "rToken";
    private static final String TEST_CLIENT_SECRET = "cSecret";
    private static final String TEST_CLIENT_ID = "cId";
    private static final String TEST_ENDPOINT = "https://www.amazon.com/lwa";
    private static final String TEST_SCOPE_1 = SCOPE_NOTIFICATIONS_API;
    private static final String TEST_SCOPE_2 = SCOPE_MIGRATION_API;
    private static final String SELLER_TYPE_SELLER = "seller";
    private static final String SELLER_TYPE_SELLERLESS = "sellerless";
    private static final MediaType EXPECTED_MEDIA_TYPE = MediaType.parse("application/json; charset=utf-8");


    private Request request;
    private static LWAAuthorizationSigner underTestSeller;
    private static LWAAuthorizationSigner underTestSellerless;
    
    @Mock
    private OkHttpClient mockOkHttpClient;

    @Mock
    private Call mockCall;

    static {

        underTestSeller = new LWAAuthorizationSigner(LWAAuthorizationCredentials.builder()
                .clientId(TEST_CLIENT_ID)
                .clientSecret(TEST_CLIENT_SECRET)
                .refreshToken(TEST_REFRESH_TOKEN)
                .endpoint(TEST_ENDPOINT)
                .build());

        underTestSellerless = new LWAAuthorizationSigner(LWAAuthorizationCredentials.builder()
                .clientId(TEST_CLIENT_ID)
                .clientSecret(TEST_CLIENT_SECRET)
                .withScopes(TEST_SCOPE_1, TEST_SCOPE_2)
                .endpoint(TEST_ENDPOINT)
                .build());
    }

    @Before @BeforeEach
    public void init() {
        request = new Request.Builder()
                .url("https://www.amazon.com/api")
                .build();
        MockitoAnnotations.initMocks(this);

    }

    public static Stream<Arguments> lwaAuthSigner(){

        return Stream.of(
                Arguments.of(SELLER_TYPE_SELLER, underTestSeller),
                Arguments.of(SELLER_TYPE_SELLERLESS, underTestSellerless)
        );
    }

    @ParameterizedTest
    @MethodSource("lwaAuthSigner")
    public void initializeLWAClientWithConfiguredEndpoint(String sellerType, LWAAuthorizationSigner testAuthSigner) {
        assertEquals(TEST_ENDPOINT, testAuthSigner.getLwaClient().getEndpoint());
    }

    @ParameterizedTest
    @MethodSource("lwaAuthSigner")
    public void requestLWAAccessTokenFromConfiguration(String sellerType, LWAAuthorizationSigner testAuthSigner) {
        LWAClient mockLWAClient = mock(LWAClient.class);
        ArgumentCaptor<LWAAccessTokenRequestMeta> lwaAccessTokenRequestMetaArgumentCaptor = ArgumentCaptor.forClass(LWAAccessTokenRequestMeta.class);

        when(mockLWAClient.getAccessToken(lwaAccessTokenRequestMetaArgumentCaptor.capture()))
                .thenReturn("foo");

        testAuthSigner.setLwaClient(mockLWAClient);
        testAuthSigner.sign(request);

        LWAAccessTokenRequestMeta actualLWAAccessTokenRequestMeta = lwaAccessTokenRequestMetaArgumentCaptor.getValue();
        assertEquals(TEST_REFRESH_TOKEN, actualLWAAccessTokenRequestMeta.getRefreshToken());
        assertEquals(TEST_CLIENT_SECRET, actualLWAAccessTokenRequestMeta.getClientSecret());
        assertEquals(TEST_CLIENT_ID, actualLWAAccessTokenRequestMeta.getClientId());

        if(sellerType.equals(SELLER_TYPE_SELLER)){
            Assert.assertTrue(actualLWAAccessTokenRequestMeta.getScopes().getScopes().isEmpty());
            assertEquals("refresh_token", actualLWAAccessTokenRequestMeta.getGrantType());
        }
        else if (sellerType.equals(SELLER_TYPE_SELLERLESS)){
            assertEquals(new HashSet<String>(Arrays.asList(TEST_SCOPE_1, TEST_SCOPE_2)), actualLWAAccessTokenRequestMeta.getScopes().getScopes());
            assertEquals("client_credentials", actualLWAAccessTokenRequestMeta.getGrantType());
        }

    }

    @ParameterizedTest
    @MethodSource("lwaAuthSigner")
    public void returnSignedRequestWithAccessTokenFromLWAClient(String sellerType, LWAAuthorizationSigner testAuthSigner) {
        LWAClient mockLWAClient = mock(LWAClient.class);

        when(mockLWAClient.getAccessToken(any(LWAAccessTokenRequestMeta.class)))
                .thenReturn("Azta|Foo");

        testAuthSigner.setLwaClient(mockLWAClient);
        Request actualSignedRequest = testAuthSigner.sign(request);

        assertEquals("Azta|Foo", actualSignedRequest.header("x-amz-access-token"));
    }

    @ParameterizedTest
    @MethodSource("lwaAuthSigner")
    public void originalRequestIsImmutable(String sellerType, LWAAuthorizationSigner testAuthSigner) {
        LWAClient mockLWAClient = mock(LWAClient.class);

        when(mockLWAClient.getAccessToken(any(LWAAccessTokenRequestMeta.class)))
                .thenReturn("Azta|Foo");

        testAuthSigner.setLwaClient(mockLWAClient);
        assertNotSame(request, testAuthSigner.sign(request));
    }
    
    @Test
    public void returnSignedRequestWithAccessTokenFromLWACache() throws IOException { 
        LWAClient testLWAClient = new LWAClient(TEST_ENDPOINT);
        testLWAClient.setOkHttpClient(mockOkHttpClient);
        
        when(mockOkHttpClient.newCall(any(Request.class)))
        .thenReturn(mockCall);
        when(mockCall.execute())     
        .thenReturn(buildResponse(200, "Azta|foo", "120"))
        .thenReturn(buildResponse(200, "Azta|foo1", "1"));

        LWAAccessTokenCache testLWACache = new LWAAccessTokenCacheImpl();        
        LWAAuthorizationSigner testlwaSigner = new LWAAuthorizationSigner(LWAAuthorizationCredentials.builder()
                .clientId(TEST_CLIENT_ID)
                .clientSecret(TEST_CLIENT_SECRET)
                .refreshToken(TEST_REFRESH_TOKEN)
                .endpoint(TEST_ENDPOINT)
                .build() , testLWACache );  
        
        testlwaSigner.setLwaClient(testLWAClient);
        testLWAClient.setLWAAccessTokenCache(testLWACache);
        Request actualSignedRequest = testlwaSigner.sign(request);
        Request actualSignedSecondRequest = testlwaSigner.sign(request);

        assertEquals("Azta|foo", actualSignedSecondRequest.header("x-amz-access-token"));
    }
    
    private static Response buildResponse(int code, String accessToken, String expiryInSeconds) {
        ResponseBody responseBody = ResponseBody.create(EXPECTED_MEDIA_TYPE,
                String.format("{%s:%s,%s:%s}", "access_token", accessToken, "expires_in", expiryInSeconds));

        return new Response.Builder()
                .request(new Request.Builder().url(TEST_ENDPOINT).build())
                .code(code)
                .body(responseBody)
                .protocol(Protocol.HTTP_1_1)
                .message("OK")
                .build();
    }
}

File: clients/sellingpartner-api-aa-php/tests/authandauth/LWAAuthorizationSignerTest.php
<?php

namespace SpApi\Test\AuthAndAuth;

use GuzzleHttp\Client;
use GuzzleHttp\Handler\MockHandler;
use GuzzleHttp\HandlerStack;
use GuzzleHttp\Psr7\Request;
use GuzzleHttp\Psr7\Response;
use SpApi\AuthAndAuth\LWAAccessTokenCache;
use SpApi\AuthAndAuth\LWAAccessTokenRequestMeta;
use SpApi\AuthAndAuth\LWAAuthorizationCredentials;
use SpApi\AuthAndAuth\LWAAuthorizationSigner;
use SpApi\AuthAndAuth\LWAClient;
use SpApi\AuthAndAuth\ScopeConstants;
use PHPUnit\Framework\TestCase;
use Psr\Http\Message\ResponseInterface;

final class LWAAuthorizationSignerTest extends TestCase
{
    private const TEST_REFRESH_TOKEN = "rToken";
    private const TEST_CLIENT_SECRET = "cSecret";
    private const TEST_CLIENT_ID = "cId";
    private const TEST_ENDPOINT = "https://www.amazon.com/lwa";
    private const SCOPE_NOTIFICATIONS_API = ScopeConstants::SCOPE_NOTIFICATIONS_API;
    private const SCOPE_MIGRATION_API = ScopeConstants::SCOPE_MIGRATION_API;
    private const SELLER_TYPE_SELLER = "seller";
    private const SELLER_TYPE_SELLERLESS = "sellerless";

    private Request $request;

    private static ?LWAAuthorizationSigner $underTestSeller = null;
    private static ?LWAAuthorizationSigner $underTestSellerless = null;

    public static function setUpBeforeClass(): void
    {
        self::$underTestSeller = new LWAAuthorizationSigner(
            new LWAAuthorizationCredentials([
                "clientId" => self::TEST_CLIENT_ID,
                "clientSecret" => self::TEST_CLIENT_SECRET,
                "refreshToken" => self::TEST_REFRESH_TOKEN,
                "endpoint" => self::TEST_ENDPOINT
            ])
        );

        self::$underTestSellerless = new LWAAuthorizationSigner(
            new LWAAuthorizationCredentials([
                "clientId" => self::TEST_CLIENT_ID,
                "clientSecret" => self::TEST_CLIENT_SECRET,
                "scopes" => [
                    self::SCOPE_NOTIFICATIONS_API,
                    self::SCOPE_MIGRATION_API
                ],
                "endpoint" => self::TEST_ENDPOINT
            ])
        );
    }

    protected function setUp(): void
    {
        $this->request = new Request("GET", "https://www.amazon.com/api");
    }

    public function lwaSignerProvider(): array
    {
        return [
            [self::SELLER_TYPE_SELLER, &self::$underTestSeller],
            [self::SELLER_TYPE_SELLERLESS, &self::$underTestSellerless]
        ];
    }

    /**
     * @dataProvider lwaSignerProvider
     */
    public function testInitializeLWAClientWithConfiguredEndpoint(
        string $sellerType,
        LWAAuthorizationSigner $testAuthSigner
    ): void {
        $this->assertSame(
            self::TEST_ENDPOINT,
            $testAuthSigner->getLwaClient()->getEndpoint()
        );
    }

    /**
     * @dataProvider lwaSignerProvider
     */
    public function testRequestLWAAccessTokenFromConfiguration(
        string $sellerType,
        LWAAuthorizationSigner $testAuthSigner
    ) {
        $mockLWAClient = $this->createMock(LWAClient::class);
        $mockLWAClient
            ->method("getAccessToken")
            ->willReturnCallback(function (LWAAccessTokenRequestMeta &$lwaAccessTokenRequestMeta) use ($sellerType) {
                $this->assertEquals(
                    self::TEST_CLIENT_SECRET,
                    $lwaAccessTokenRequestMeta->getClientSecret()
                );
                $this->assertEquals(
                    self::TEST_CLIENT_ID,
                    $lwaAccessTokenRequestMeta->getClientId()
                );

                if ($sellerType == self::SELLER_TYPE_SELLER) {
                    $this->assertEmpty($lwaAccessTokenRequestMeta->getScopes());
                    $this->assertEquals("refresh_token", $lwaAccessTokenRequestMeta->getGrantType());
                    $this->assertEquals(
                        self::TEST_REFRESH_TOKEN,
                        $lwaAccessTokenRequestMeta->getRefreshToken()
                    );
                } elseif ($sellerType == self::SELLER_TYPE_SELLERLESS) {
                    $this->assertEquals(
                        [
                            self::SCOPE_NOTIFICATIONS_API,
                            self::SCOPE_MIGRATION_API
                        ],
                        $lwaAccessTokenRequestMeta->getScopes()
                    );
                    $this->assertEquals("client_credentials", $lwaAccessTokenRequestMeta->getGrantType());
                    $this->assertNull($lwaAccessTokenRequestMeta->getRefreshToken());
                }
                return "foo";
            });

        $testAuthSigner->setLwaClient($mockLWAClient);
        $testAuthSigner->sign($this->request);
    }

    /**
     * @dataProvider lwaSignerProvider
     */
    public function testReturnSignedRequestWithAccessTokenFromLWAClient(
        string $sellerType,
        LWAAuthorizationSigner $testAuthSigner
    ) {
        $mockLWAClient = $this->createMock(LWAClient::class);

        $mockLWAClient->method("getAccessToken")->willReturn("Azta|Foo");

        $testAuthSigner->setLwaClient($mockLWAClient);
        $actualSignedRequest = $testAuthSigner->sign($this->request);
        $this->assertEquals("Azta|Foo", $actualSignedRequest->getHeader("x-amz-access-token")[0]);
    }

    /**
     * @dataProvider lwaSignerProvider
     */
    public function testOriginalRequestIsImmutable(string $sellerType, LWAAuthorizationSigner $testAuthSigner)
    {
        $mockLWAClient = $this->createMock(LWAClient::class);

        $mockLWAClient->method("getAccessToken")->willReturn("Azta|Foo");

        $testAuthSigner->setLwaClient($mockLWAClient);
        $this->assertNotSame($this->request, $testAuthSigner->sign($this->request));
    }

    public function testReturnSignedRequestWithAccessTokenFromLWACache()
    {
        $mockHandler = new MockHandler([
            $this->buildResponse(200, "Azta|foo", "120"),
            $this->buildResponse(200, "Azta|foo1", "1")
        ]);

        $handlerStack = HandlerStack::create($mockHandler);
        $mockGuzzleClient = new Client(["handler" => $handlerStack]);

        $testLWAClient = new LWAClient(self::TEST_ENDPOINT);
        $testLWAClient->setClient($mockGuzzleClient);

        $testLWACache = new LWAAccessTokenCache();

        $testlwaSigner = new LWAAuthorizationSigner(
            new LWAAuthorizationCredentials([
                "clientId" => self::TEST_CLIENT_ID,
                "clientSecret" => self::TEST_CLIENT_SECRET,
                "refreshToken" => self::TEST_REFRESH_TOKEN,
                "endpoint" => self::TEST_ENDPOINT
            ]),
            $testLWACache
        );

        $testlwaSigner->setLwaClient($testLWAClient);
        $testLWAClient->setLWAAccessTokenCache($testLWACache);
        $actualSignedRequest = $testlwaSigner->sign($this->request);
        $actualSignedSecondRequest = $testlwaSigner->sign($this->request);

        $this->assertEquals("Azta|foo", $actualSignedSecondRequest->getHeader("x-amz-access-token")[0]);
    }

    private function buildResponse(int $code, string $accessToken, string $expiryInSeconds = "3600"): ResponseInterface
    {
        $contentHeader = [
            "Content-Type" => "application/json",
        ];
        return new Response(
            $code,
            $contentHeader,
            json_encode(["access_token" => $accessToken, "expires_in" => $expiryInSeconds])
        );
    }
}
